File: backend/src/main/java/com/example/backend/model/Category.java
package com.example.backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(name = "danh_muc")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "ten_danh_muc", nullable = false, length = 100)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(name = "loai", nullable = false)
    private CategoryType type = CategoryType.EXPENSE;

    @Column(name = "id_cha")
    private Long parentId;

    @Column(name = "la_he_thong")
    private Boolean isSystem = false;

    @Column(name = "thu_tu_hien_thi")
    private Integer displayOrder = 0;

    @Column(name = "bieu_tuong", length = 100)
    private String icon;

    @Column(name = "mau_sac", length = 20)
    private String color;

    @Column(name = "mo_ta", columnDefinition = "TEXT")
    private String description;

    @Column(name = "ngay_tao", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "ngay_cap_nhat")
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_cha", insertable = false, updatable = false)
    private Category parent;

    @OneToMany(mappedBy = "parent", fetch = FetchType.LAZY)
    private java.util.List<Category> subCategories;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public enum CategoryType {
        INCOME, EXPENSE
    }
}

File: finpal_mobile/lib/data/models/category.dart
class Category {
  final int id;
  final String name;
  final String type; // INCOME or EXPENSE
  final int? parentId;
  final bool isSystem;
  final int displayOrder;
  final String? icon;
  final String? color;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final Category? parent;
  final List<Category>? subCategories;

  Category({
    required this.id,
    required this.name,
    required this.type,
    this.parentId,
    required this.isSystem,
    required this.displayOrder,
    this.icon,
    this.color,
    this.createdAt,
    this.updatedAt,
    this.parent,
    this.subCategories,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'] as int,
      name: json['name'] as String,
      type: json['type'] as String,
      parentId: json['parentId'] as int?,
      isSystem: json['isSystem'] as bool? ?? true,
      displayOrder: json['displayOrder'] as int? ?? 0,
      icon: json['icon'] as String?,
      color: json['color'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
      parent: json['parent'] != null
          ? Category.fromJson(json['parent'] as Map<String, dynamic>)
          : null,
      subCategories: json['subCategories'] != null
          ? (json['subCategories'] as List)
                .map((e) => Category.fromJson(e as Map<String, dynamic>))
                .toList()
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type,
      'parentId': parentId,
      'isSystem': isSystem,
      'displayOrder': displayOrder,
      'icon': icon,
      'color': color,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'parent': parent?.toJson(),
      'subCategories': subCategories?.map((e) => e.toJson()).toList(),
    };
  }
}
